Upload data.json as an artifact on each build

karx1 opened the pull request:
This PR uploads the `data.json` file as an artifact on each build. This way, we can track past pogs so we can access them later.

This PR closes #3.

karx1 committed:
Upload data.json as an artifact on each build

MattIPv4 reviewed (changes requested):
Before attempting to upload to artifacts, we should first attempt to download the previous version. We can then compare the local file to the downloaded one and see if they match, only continuing the workflow (upload & deploy) if they don't match. This will help to reduce load and compute time when we don't need to do anything.

---

The other thing here is that this doesn't really help with tracking the history of the emote, as we're uploading data.json each time, which currently only contains the most recent emote. We likely need to use jq or similar to append the latest emote data to the previously downloaded history, and then push that back up. (Artifacts aren't retained forever, so we need a rolling history in the latest artifact).

karx1 committed:
Compare and combine artifacts

karx1 committed:
use alternative downloader

This is because the official workflow doesn't support downloading from previous runs.

karx1 commented:
@MattIPv4 Now it can compare with previous artifacts an will exit if they are the same. It also, if they're not the same, will append the JSON to a different file, so we'll have the "rolling" effect of the artifacts.

MattIPv4 commented on .github/workflows/build.yml:
```diff
@@ -50,7 +50,34 @@ jobs:
         run: cp build/data.json dist/data.json
+        
+      - name: Download old artifact
+        uses: dawidd6/action-download-artifact@v2
```
Why is this needed over the default download?

MattIPv4 commented on .github/workflows/build.yml:
```diff
@@ -50,7 +50,34 @@ jobs:
+          path: build/old.json
+
+      - name: Compare both artifacts
+        run: (diff build/data.json build/old.json &&  echo "NOT OK" | grep "NOT OK") && false # will exit the workflow with an error, stopping next build steps
```
This will exit code 1 if the files match, which will cause the run to be flagged as failing. We need to cleanly stop the workflow if the files are the same (maybe using the `if` functionality of actions?).

MattIPv4 commented on .github/workflows/build.yml:
```diff
@@ -50,7 +50,34 @@ jobs:
+          # Artifact name
+          name: all.json
+          path: build/all.json
+      
```
You also need to upload data.json here so that it can be referenced in subsequent builds to check if the emote has changed?

karx1 commented:
> Why is this needed over the default download?
>

This is needed because the default download can't pull artifacts from past workflow runs. I wrote that in the commit description.


> This will exit code 1 if the files match, which will cause the run to be flagged as failing. We need to cleanly stop the workflow if the files are the same (maybe using the if functionality of actions?).

I'll look into that. Raising an exit code 0 will simply stop that individual step, but continue with the rest of the workflow.


> You also need to upload data.json here so that it can be referenced in subsequent builds to check if the emote has changed?
>

Sorry, forgot about that.

MattIPv4 commented:
>  > Why is this needed over the default download?
> 
> This is needed because the default download can't pull artifacts from past workflow runs. I wrote that in the commit description.

Ah (TIL) -- doesn't the other download also need this then to persist the history?

> > This will exit code 1 if the files match, which will cause the run to be flagged as failing. We need to cleanly stop the workflow if the files are the same (maybe using the if functionality of actions?).
> 
> I'll look into that. Raising an exit code 0 will simply stop that individual step, but continue with the rest of the workflow.

Yeah, I think you'll want to exit 0 but also set an output in the step, and then subsequent steps can use an if to check that output I believe.

Tyvm!

MattIPv4 commented:
Thinking out loud here, maybe instead of persisting the history in artifacts, we persist it in the gh-pages branch itself.

And regarding the `if` logic, perhaps we split this into two jobs: the first job fetches the 'live' emote data and the latest emote data in the artifact, outputting the data only if its new. The second job then only runs if the first job actually output something, and handles updating the history and building the website?

MattIPv4 commented:
👋 I ended up playing with this idea a load more and think I've come up with a solution that works, having the data & history tracked as JSON files in master, with a dedicated workflow to handle checking for updates, and another that builds the site.

I realise this puts to waste the work done here, but I feel its probably a more robust solution than relying on artifacts in the long-run, as git history is permanent but artifacts are not.

MattIPv4 commented:
Tyvm for taking a look at this though, definitely acted as the catalyst needed to figure out this solution 💙
